components/http-client/src/api_client.rs

```diff
@@ -191,121 +188,113 @@ impl ApiClient {
+/// IN ADDITION to any system certificates that may be available (e.g., in /etc/ssl/certs or
+/// specified by a `SSL_CERT_FILE` environment variable):
+///
+/// 1. If the `core/cacerts` Habitat package is installed locally, then use the latest release's
+///    `cacert.pem` file.
+/// 2. If there is no 'core/cacerts packages, then a copy of `cacert.pem` will be written in an SSL
+///    cache directory (by default `/hab/cache/ssl` for a root user and `$HOME/.hab/cache/ssl` for
+///    a non-root user) and this will be used. The contents of this file will be inlined in this
+///    crate at build time as a fallback, which means that if the program using this code is
+///    operating in a minimal environment which may not contain any system certificates, it can
+///    still operate.
+/// 3. Other certs files (for example self-signed certs) that are found in the SSL cache directory
+///    will also get loaded into the root certs list. Both PEM and DER formats are supported (the
+///    extensions should be '.pem' or '.der' respectively)
+fn certificates(fs_root_path: Option<&Path>) -> Result<Vec<Certificate>> {
+    let mut certificates = Vec::new();
+    let cacerts_ident = PackageIdent::from_str(CACERTS_PKG_IDENT)?;
+
+    if let Ok(pkg_install) = PackageInstall::load(&cacerts_ident, fs_root_path) {
+        let pkg_certs = pkg_install.installed_path().join("ssl/cert.pem");
+        debug!("Found installed certs: {}", pkg_certs.display());
+        let cert = cert_from_file(&pkg_certs)?;
+        certificates.push(cert);
+    } else {
+        debug!("No installed cacerts package found");
+
+        let cached_certs = cache_ssl_path(fs_root_path).join("cert.pem");
+        if !cached_certs.exists() {
+            fs::create_dir_all(cache_ssl_path(fs_root_path))?;
+            debug!("Creating cached cacert.pem: {}", cached_certs.display());
+            let mut file = File::create(&cached_certs)?;
+            file.write_all(CACERT_PEM.as_bytes())?;
+        }
     }
 
-    Ok(conn.build())
+    if let Ok(entries) = fs::read_dir(cache_ssl_path(fs_root_path)) {
+        for entry in entries {
+            let entry = entry?;
+            let path = entry.path();
+
+            if let Some(ext) = path.extension() {
+                if path.is_file() && ((ext == "pem") || (ext == "der")) {
+                    debug!("Found cached cert: {}", path.display());
+                    let cert = cert_from_file(&path)?;
+                    certificates.push(cert);
+                }
+            }
+        }
+    }
+
+    Ok(certificates)
+}
+
+fn cert_from_file(file_path: &PathBuf) -> Result<Certificate> {
+    let mut buf = Vec::new();
+    File::open(file_path)?.read_to_end(&mut buf)?;
+
+    let ext = file_path.extension().unwrap(); // unwrap Ok
+    if ext == "pem" {
+        Certificate::from_pem(&buf).map_err(Error::ReqwestError)
+    } else {
+        Certificate::from_der(&buf).map_err(Error::ReqwestError)
+    }
```

Comment:
To reduce the duplication here, you could do something like this:

```rust
let fun = if ext == "pem" {
    Certificate::from_pem
} else {
    Certificate::from_der
};

fun(&buf).map_err(Error::ReqwestError)
```

Reply:
Yep, going with Jon's suggestion on this...